client/state/routing/middleware.js

```diff
@@ -13,6 +13,11 @@ import { ROUTE_SET } from 'state/action-types';
 
 const debug = debugFactory( 'calypso:restore-last-location' );
 const LAST_PATH = 'last_path';
+const ALLOWED_PATHS = /^\/(stats|plans|view|posts|pages|media|types|themes|sharing|people|plugins|domains)/i;
```

Comment:
Yeah. It might be a challenge as the URL structure is not alway that simple. 

Also, I'm little scared that unless we don't implement the whitelist as part of route declaration, it might easily fall out of sync. This version seems dumb enough to be OK but higher configurability level might be a problem.